target/linux/ath79/image/nand.mk

```diff
@@ -74,6 +74,37 @@ define Device/glinet_gl-ar300m-nor
 endef
 TARGET_DEVICES += glinet_gl-ar300m-nor
 
+define Device/glinet_gl-ar750s-common
+  ATH_SOC := qca9563
+  DEVICE_VENDOR := GL.iNet
+  DEVICE_MODEL := GL-AR750S
+  DEVICE_PACKAGES := kmod-ath10k-ct ath10k-firmware-qca9887-ct \
+			kmod-usb2 kmod-usb-storage block-mount
+  KERNEL_SIZE := 2048k
```

Comment:
Could this be increased to 4M? A 2M kernel partition won't be enough after several more kernel bumps, and if we change this after it's not enough, sysupgrade between firmwares will break.

Reply:
I agree that kernel size is a significant concern. See, for example, https://forum.openwrt.org/t/porting-wndr4300-to-ath79/16006/32?u=jeff

To be able to increase the kernel size will require confirming that the OEM U-Boot can handle kernels of that size properly. As current source for the two, U-Boot implementations are not reliably available to me, I have asked GL.iNet to comment.

https://forum.gl-inet.com/t/nand-kernel-size-and-u-boot-preventing-obsolescence/8730?u=jeffsf

Reply:
**TL;DR**

* No _OpenWrt_ change appears needed for the GL-AR300M series to expand the kernel size in the future through sysupgrade, however

* Formally increasing the kernel size limit for the GL-AR300M probably should wait for GL.iNet to confirm that all U-Boot functionality is compatible with kernels over 2 MB in size

* Increasing the kernel size for the GL-AR750S likely requires replacement of the OEM U-Boot and should wait for GL.iNet to update the OEM U-Boot


In this discussion, I’ll first look at sysupgrade, then look into the OEM U-Boot, based on their public repositories.

**sysupgrade**

I've been staring at the code for ath79, as well as for ar71xx, both that of OpenWrt and of GL.iNet. 

It appears that a "pure NOR" case is "transparently" handled by `default_do_upgrade()` which “streams” a combined U-Boot header, kernel, ROM, and JFFS2 file system to the combined firmware partition. The creation of the MTD `kernel` partition is handled dynamically during early boot.

It also appears that a “pure NAND” case is also handled transparently as `nand_upgrade_prepare_ubi()` destroys the existing kernel UBI volume and creates a new one of an appropriate size to accommodate the kernel in the tar.

This covers the GL-AR300M series, as they do not have a variant where the kernel is on NOR flash and the file systems are on NAND. However, as discussed below, it is possible that a kernel larger than 2 MB may break U-Boot flashing, even if U-Boot is able to load and boot the kernel

The GL-AR750S presently always has a NOR-resident kernel and the troublesome operation can be seen at 

https://github.com/openwrt/openwrt/blob/master/package/base-files/files/lib/upgrade/nand.sh#L260

```
	[ "$kernel_length" != 0 -a -n "$kernel_mtd" ] && {
		tar xf $tar_file ${board_dir}/kernel -O | mtd write - $CI_KERNPART
	}
```

In this situation, the kernel is extracted from the tar and written to the `kernel` MTD partition on NOR flash. AFAIK, dynamic resizing of MTD partitions once the kernel is “up and running” is not possible.

Since the “pure NOR” case should work properly (assuming that the U-Boot code can boot a kernel over 2 MB in size), one possible upgrade path would be to flash a “pure NOR” image with the increased `nor_kernel` partition size, then cross-flash to the NOR-NAND variant.

**U-Boot**

My U-Boot expertise is minimal, so my reading of the code and understanding of the behavior of U-Boot is inexpert. My apologies if I have misinterpreted what I see.

**U-Boot — GL-AR750S**

https://github.com/gl-inet/uboot-gl-ar750-ar750s-x750

It appears that increasing the kernel size over 2 MB will “break” the ability to flash through U-Boot

From `src/include/configs/board956x.h`

```
49b55c13 (Luochongjun 2018-05-25 15:19:46 +0800  74) #elif defined(CONFIG_FOR_GL_BOARD)                                 
64a006b2 (deng        2018-03-08 21:39:57 -0800  75)    #define WEBFAILSAFE_UPLOAD_KERNEL_ADDRESS                       WEBFAILSAFE_UPLOAD_UBOOT_ADDRESS + 0x60000
1051557b (deng        2018-04-11 10:24:11 -0700  76)    #define WEBFAILSAFE_UPLOAD_NAND_KERNEL_SIZE                     0x200000
a5f2190f (deng        2018-03-07 22:20:48 -0800  77) #else
```

This seems to indicate that a kernel over 2 MB would be rejected, at least through the U-Boot GUI

```
1051557b (deng        2018-04-11 10:24:11 -0700 161) #define VAR_KERNEL_SIZE 0x00200000
```

From the running device
```
kernelsize=0x00200000

lf=if ping $serverip; then tftp $loadaddr $firmware_nand_name && erase $firmware_addr +$kernelsize && nand erase && cp.b $fileaddr $firmware_addr $kernelsize && nand write $rootfs_addr 0 $rootfs_size && echo OK!; else ERROR! Server not reachable!; fi
```

This seems to indicate that flashing a kernel over 2 MB through the U-Boot command line would result in unexpected behavior.

While the on-flash U-Boot environment could conceivably be modified, I have seen devices “reset” their U-Boot environments to default values. As a result, I don’t believe such a course of action to be robust.

**U-Boot — GL-AR300M**

https://github.com/gl-inet/uboot-ar300m

There is not a similar parameter for kernel-size that I could find in `src/include/configs/board953x.h`. This seems to be due to the fact that U-Boot is either flashing a single, self-contained image to the NOR firmware partition or to the NAND flash.

Looking at `./src/httpd/httpd.c` (lines 225-231) shows that the file-size check for firmware is applied to 

```
FLASH_SIZE -  WEBFAILSAFE_UPLOAD_LIMITED_AREA_IN_BYTES
```
which appears to be set at ``src/include/configs/board953x.h#90` to 192 kB
```
#define WEBFAILSAFE_UPLOAD_LIMITED_AREA_IN_BYTES        (192 * 1024)
```

The firmware-related, U-Boot commands visible from the my running GL-AR300M’s U-Boot environment appear to be more amenable to larger kernels, assuming that `$filesize` and `$web_filesize` are dynamically generated from the actual upload. Neither of these two variables appeared set on my running device when starting U-Boot.

```
dlf=if ping 192.168.1.2; then echo ok; elif ping 192.168.1.2; then echo ok; elif ping 192.168.1.2; then echo ok; elif echo ping 192.168.1.2; then echo ok; elif echo ping 192.168.1.2; then echo ok; else echo ping finally failed; fi; tftp 0x81000000 openwrt-gl-ar300m-ubi.img && nand erase && nand write $fileaddr 0 $filesize

wlf=if nand bad; then nand erase && nand write $web_fileaddr 0 $web_filesize; else erase 0x9f050000 +e30000 && cp.b $web_fileaddr 0x9f050000 $web_filesize; fi; erase 0x9f040000 +0x10000

rlf=if ping 192.168.1.2; then tftp 0x80060000 ${dir}openwrt-gl-ar300m.bin && erase 0x9f050000 +$filesize && cp.b $fileaddr 0x9f050000 $filesize; else echo ping 192.168.1.2 failed; fi

lf=if ping 192.168.1.2; then tftp 0x80060000 ${dir}openwrt-gl-ar300m.bin && erase 0x9f050000 +$filesize && cp.b $fileaddr 0x9f050000 $filesize; if nand bad; then run dlf; fi; else echo ping 192.168.1.2 failed; fi
```

Reply:
Thanks for pointing out the error in my thinking around the NAND-based kernel partition. I've struck-out my mistaken thoughts from above to keep the conversation clear to any that might find this later.

As much as I want to be done with this, doing it right is pushing me to explore what it would take to partition for 4 MB kernel at this time. I also know that it would likely be me that ends up doing the work in the future. 

My current thinking is to enlarge the partitions at this time but keep the kernel-size limit at 2 MB. At first glance, this seems as though it would
* Provide smooth upgrades in the future, when the kernel-size limit is increased
* Continue to allow the current, GL-AR750S U-Boot to be able to flash images
* Avoid any potential, unrevealed issues with the current U-Boot for either device with a kernel over 2 MB in size
* Provide "20.01 release" as a "stepping stone" from older releases to ones with a kernel over 2 MB

I've got of flash and upgrade paths to work through, first in my head, then with live hardware.

There's also the complication of the existing `glinet,gl-ar300m-nor` and `glinet,gl-ar300m-nand` sitting on the openwrt-19.07 branch. I'm already tempted to ask to get `glinet,gl-ar300m-nor` changed to `glinet,gl-ar300m16` prior to 19.07-RC so that there is clear disambiguation between those _release_ OpenWrt builds that have a NAND-aware kernel and appropriate partitioning and would be compatible with such a future upgrade

* glinet,gl-ar300m-nor
* glinet,gl-ar300m-nand

and those that do not 

* gl-ar300m
* glinet,gl-ar300m16
* glinet,gl-ar300m-lite

Any thoughts on that?

_Bottom line; I'm considering the impact of increasing to a 4 MB partition for the kernel on both devices, prior to merge, but don't know enough yet to firmly commit to it._

Reply:
rc1 is tagged. How about disabling ar300m-nor and ar300m-nand in openwrt-19.07 (and removing the images build for rc1) until this is sorted out? (That's just a personal thought, a do not know whether this will be accepted by the others.)

Reply:
@adrianschmutzler 

That would certainly prevent a lot of future pain, as there haven't been any _release_ firmware  for those devices on the ath79 platform.

I've tried to get the apparently non-functional gl-ar300m-nand dropped in the past, "crickets ...". Now that I can risk making a complete mess of my GL-AR300M, I was planning on "proving" that it doesn't run.

I was hoping to hold off until this was merged, to further strengthen the arguments, but yesterday's tagging of RC1 puts me in the hot seat again.

Reply:
I've removed glinet,gl-ar300m-nor, glinet,gl-ar300m-nand and glinet,gl-ar750s from TARGET_DEVICES in ath79 openwrt-19.07:
https://github.com/openwrt/openwrt/commit/5fc63ec8a496066d101de41ae4f820c5c4e39c04

This will prevent images from being built, and jow has removed already built images for 19.07.0-rc1. Code is still there, but if someone had wanted to build it, he would have had enough time already.

Reply:
Thank you!!!!!

Not having _formal release_ firmware with those board names makes things so much easier going forward. 

Yes, there are certain "nut cases" that have built and flashed probably hundreds of images from that ath79 source (though perhaps just one to that extreme). At least in my opinion, working with development or unreleased branches does _not_ come with the same "guarantees" as does flashing a formal release image.

Reply:
Well, the biggest problem in my opinion will be downstream projects building based on openwrt-19.07. However, those will still have to patch the TARGET_DEVICES comment away ...

Reply:
GL.iNet is aware of the previous lack of SPI-NAND support on the ath79 target, as well as this PR. 

From what I can see of their public repository, they are still building firmware for their new, QCA-based devices with SPI-NAND on the deprecated ar71xx driver. See, for example, from August 13, https://github.com/gl-inet/openwrt/commit/cfc85fc80e914a2808f7b19d71c47db24c471cd7

I'll certainly ensure they're aware of the progress here. 

I originally developed this prior to the `openwrt-19.07` branch and the subsequent change on `master` to Linux 4.19. Checking my old branches, I see
```
commit 61883e9cee
Author: Jeff Kletsky <git-commits@allycomm.com>
Date:   Tue May 14 09:47:40 2019 -0700

    ath79: Prepare NAND subtarget for upstream support of SPI NAND
    
    Linux 4.19 supplies the upstream spi-nand framework,
    permitting porting and support of boards with SPI NAND.
    
      * Adjusted nand/target.mk to set KERNEL_PATCHVER:=4.19
```
which was producing functional images.

It then becomes a commercial decision if GL.iNet chooses to continue to use the deprecated driver at this time, or to backport the necessary commits for a NAND-enabled ath79 target on forthcoming releases of theirs, before they rebase on "20.01", or whatever the next, formal release is called.

Reply:
Since openwrt-19.07 is kernel 4.14 only, backporting ath79 nand from master doesn't sound very desirable for me. I'd rather say 19.07 is a ar71xx release (as planned with ath79 being moved out of source-only very late), and 20.xx will be an ath79 release then.